### PR TITLE
Improve page drag and drop experience

### PR DESCRIPTION
**Pull request by mejo-**
If a page doesn't have a subpage yet (or it's not expanded), dragging another page underneath it doesn't create a new subpage list. This behaviour created too much UX confusion at selecting the target level when dragging a page at the end of expanded subpages.

Instead, dragging a page *over* a page without subpages will highlight the dragover page and dropping will turn the dragged page into a subpage of the dragover page.

For this, in addition to the Sortable.js logic, we add custom drag and drop handlers in Item.vue.

Also, when a page moving operation is pending (waiting for the backend response), moving/sorting is disabled now to prevent race conditions.

Related: #423 

Signed-off-by: Jonas <jonas@freesources.org>

### 🚧 TODO

- [ ] Let users try out the new behaviour and see whether it works for them

### 🏁 Checklist

- [x] Code is properly formatted (`npm run lint` / `npm run stylelint` / `composer run cs:check`)
- [x] [Sign-off message](https://probot.github.io/apps/dco/) is added to all commits

### CODE REVIEW

**Comment by mejo-** [comment]
Here's a zip containing pages for a test scenario. Unpack it into the folder of a Collective and try to nest and order the pages according to their naming.

[Test sorting pages.zip](https://github.com/nextcloud/collectives/files/10448136/Test.sorting.pages.zip)

**Comment by mejo-** [comment]
Notes from first test:

* When dragging a child over its parent, visual feedback is good, but dropping shouldn't trigger a move.
* When dragging a child over its parent and releasing, sometimes it gets moved above/below the parent. That's a bug.
* With page structure A, A1, B, B1 and all expanded with seamed easy to trigger the bug where move doesn't happen (#423) when pulling B above A.

**Comment by julien-nc** [comment]
@mejo- I tried to reproduce the bugs we saw together, all good as far as I can see after playing with it a few minutes.

Spotted another bug :grin:
I can consistently reproduce it by moving a page to depth 3, there is a network error but we can see it worked after reloading the page. Same when trying to get the page out of depth 3.

https://user-images.githubusercontent.com/11291457/214818176-9c9653cd-6417-4f70-bd95-7bd81934f395.mp4

Server log when moving the page to depth 3 (inside a depth 2):
```
{"reqId":"uGa7r2iOlEO8BuweW0UM","level":3,"time":"2023-01-26T10:52:37+00:00","remoteAddr":"127.0.0.1","user":"julien","app":"collectives","method":"PUT","url":"/dev/server26/index.php/apps/collectives/_api/1/_pages/parent/14461/page/14463","message":"Collectives App Error: OCA\\Collectives\\Service\\SubpageOrderService::add(): Argument #1 ($subpageOrder) must be of type string, null given, called in /datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Service/PageService.php on line 802","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/109.0","version":"26.0.0.2","exception":{"Exception":"TypeError","Message":"OCA\\Collectives\\Service\\SubpageOrderService::add(): Argument #1 ($subpageOrder) must be of type string, null given, called in /datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Service/PageService.php on line 802","Code":0,"Trace":[{"file":"/datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Service/PageService.php","line":802,"function":"add","class":"OCA\\Collectives\\Service\\SubpageOrderService","type":"::"},{"file":"/datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Service/PageService.php","line":707,"function":"addToSubpageOrder","class":"OCA\\Collectives\\Service\\PageService","type":"->"},{"file":"/datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Controller/PageController.php","line":125,"function":"rename","class":"OCA\\Collectives\\Service\\PageService","type":"->"},{"file":"/datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Controller/ErrorHelper.php","line":25,"function":"OCA\\Collectives\\Controller\\{closure}","class":"OCA\\Collectives\\Controller\\PageController","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"/datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Controller/PageController.php","line":129,"function":"handleErrorResponse","class":"OCA\\Collectives\\Controller\\PageController","type":"->"},{"file":"/datadisk/tera/julien/html/dev/server26/lib/private/AppFramework/Http/Dispatcher.php","line":230,"function":"rename","class":"OCA\\Collectives\\Controller\\PageController","type":"->"},{"file":"/datadisk/tera/julien/html/dev/server26/lib/private/AppFramework/Http/Dispatcher.php","line":137,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/datadisk/tera/julien/html/dev/server26/lib/private/AppFramework/App.php","line":171,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/datadisk/tera/julien/html/dev/server26/lib/private/Route/Router.php","line":298,"function":"main","class":"OC\\AppFramework\\App","type":"::"},{"file":"/datadisk/tera/julien/html/dev/server26/lib/base.php","line":1050,"function":"match","class":"OC\\Route\\Router","type":"->"},{"file":"/datadisk/tera/julien/html/dev/server26/index.php","line":36,"function":"handleRequest","class":"OC","type":"::"}],"File":"/datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Service/SubpageOrderService.php","Line":65,"message":"Collectives App Error: OCA\\Collectives\\Service\\SubpageOrderService::add(): Argument #1 ($subpageOrder) must be of type string, null given, called in /datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Service/PageService.php on line 802","exception":{},"CustomMessage":"Collectives App Error: OCA\\Collectives\\Service\\SubpageOrderService::add(): Argument #1 ($subpageOrder) must be of type string, null given, called in /datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Service/PageService.php on line 802"}}
```
and, after reload, when getting it out:
```
{"reqId":"Nltx6hwdB1KGhR9yS1Ib","level":3,"time":"2023-01-26T10:52:50+00:00","remoteAddr":"127.0.0.1","user":"julien","app":"collectives","method":"PUT","url":"/dev/server26/index.php/apps/collectives/_api/1/_pages/parent/14462/page/14463","message":"Collectives App Error: OCA\\Collectives\\Service\\SubpageOrderService::remove(): Argument #1 ($subpageOrder) must be of type string, null given, called in /datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Service/PageService.php on line 823","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/109.0","version":"26.0.0.2","exception":{"Exception":"TypeError","Message":"OCA\\Collectives\\Service\\SubpageOrderService::remove(): Argument #1 ($subpageOrder) must be of type string, null given, called in /datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Service/PageService.php on line 823","Code":0,"Trace":[{"file":"/datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Service/PageService.php","line":823,"function":"remove","class":"OCA\\Collectives\\Service\\SubpageOrderService","type":"::"},{"file":"/datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Service/PageService.php","line":706,"function":"removeFromSubpageOrder","class":"OCA\\Collectives\\Service\\PageService","type":"->"},{"file":"/datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Controller/PageController.php","line":125,"function":"rename","class":"OCA\\Collectives\\Service\\PageService","type":"->"},{"file":"/datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Controller/ErrorHelper.php","line":25,"function":"OCA\\Collectives\\Controller\\{closure}","class":"OCA\\Collectives\\Controller\\PageController","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"/datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Controller/PageController.php","line":129,"function":"handleErrorResponse","class":"OCA\\Collectives\\Controller\\PageController","type":"->"},{"file":"/datadisk/tera/julien/html/dev/server26/lib/private/AppFramework/Http/Dispatcher.php","line":230,"function":"rename","class":"OCA\\Collectives\\Controller\\PageController","type":"->"},{"file":"/datadisk/tera/julien/html/dev/server26/lib/private/AppFramework/Http/Dispatcher.php","line":137,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/datadisk/tera/julien/html/dev/server26/lib/private/AppFramework/App.php","line":171,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/datadisk/tera/julien/html/dev/server26/lib/private/Route/Router.php","line":298,"function":"main","class":"OC\\AppFramework\\App","type":"::"},{"file":"/datadisk/tera/julien/html/dev/server26/lib/base.php","line":1050,"function":"match","class":"OC\\Route\\Router","type":"->"},{"file":"/datadisk/tera/julien/html/dev/server26/index.php","line":36,"function":"handleRequest","class":"OC","type":"::"}],"File":"/datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Service/SubpageOrderService.php","Line":86,"message":"Collectives App Error: OCA\\Collectives\\Service\\SubpageOrderService::remove(): Argument #1 ($subpageOrder) must be of type string, null given, called in /datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Service/PageService.php on line 823","exception":{},"CustomMessage":"Collectives App Error: OCA\\Collectives\\Service\\SubpageOrderService::remove(): Argument #1 ($subpageOrder) must be of type string, null given, called in /datadisk/tera/julien/html/dev/server26/apps/collectives/lib/Service/PageService.php on line 823"}}
```

**Comment by mejo-** [comment]
> Spotted another bug grin I can consistently reproduce it by moving a page to depth 3, there is a network error but we can see it worked after reloading the page. Same when trying to get the page out of depth 3.

This was a bug I introduced when implementing `subpageOrder` updates in the backend. Should be fixed now. Mind giving it another try @julien-nc? Also @max-nextcloud since you already did quite some testing of it :wink:

File: lib/Service/SubpageOrderService.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	 * @param string|null $subpageOrder
+	 * @param int         $pageId
+	 *
+	 * @return string
+	 * @throws NotPermittedException
+	 */
+	public static function remove(?string $subpageOrder, int $pageId): string {
+		$subpageOrderArray = self::toArray($subpageOrder);
+
+		if (false !== $key = array_search($pageId, $subpageOrderArray, true)) {
```
Comment (julien-nc) [inline lib/Service/SubpageOrderService.php:92]: I always find this syntax hard to read. As you wish though :grin:

**Review by julien-nc — APPROVED** [review]
Work fine. Code looks good (and clean!) :+1:

**Pull request merged**